securitas/security/ipa.py

```diff
@@ -49,6 +49,16 @@ def group_add_member_manager(
             parse_group_management_error(data)
         return data['result']
 
+    def batch(self, methods=None):
+        """
+        Make multiple ipa calls via one remote procedure call.
+
+        :param methods: Nested Methods to execute.
+        :type methods: dict
+        """
+        data = self._request('batch', methods)
```

Comment:
For a method that fails, like this: 

    ipa.batch(methods=[{"method": "user_find", "params": [[], {"pants":None}]}])
    
the API returns the following in the result:

    {'count': 1, 'results': [{'error': "invalid 'params': Unknown option: pants", 'error_code': 3008, 'error_name': 'ConversionError', 'error_kw': {'name': 'params', 'error': 'Unknown option:pants'}}], 'messages': [{'type': 'warning', 'name': 'VersionMissing', 'message': "API Version number was not sent, forward compatibility not guaranteed. Assuming server's API version, 2.235", 'code': 13001, 'data': {'server_version': '2.235'}}]}

For a method that doesnt exist, like this:

    ipa.batch(methods=[{"method": "user_findy", "params": [[], {}]}])

we get:

    {'count': 1, 'results': [{'error': "unknown command 'user_findy'", 'error_code': 905, 'error_name': 'CommandError', 'error_kw': {'name': 'user_findy'}}], 'messages': [{'type': 'warning', 'name': 'VersionMissing', 'message': "API Version number was not sent, forward compatibility not guaranteed. Assuming server's API version, 2.235", 'code': 13001, 'data': {'server_version': '2.235'}}]}

Reply:
IS that suffient here, or do we need something more? 

Either way, added some tests to check that these errors are returned from the RPC anyways.

Reply:
If you look at how other methods work, they convert errors into subclasses of `FreeIPAError`. You can look at the `python_freeipa.exceptions.parse_*_error` methods for examples.

On the one hand I think it would be the best thing to do here, on the other hand we will probably throw away this code when we move to the non-legacy client. But since we don't know when that'll happen, I would still recommend doing the following:
- for each result in the `results` list, pass it to a `parse_batch_error` function
- that function would look for an `error` key and convert the result to an instance of `python_freeipa.exceptions.BadRequest` (ideally a relevant subclass like they do in `python_freeipa.exceptions.parse_error` but they don't have those codes in their dictionary and that's much harder to tweak, so let's settle for `BadRequest` for now).
- return the modified list
- add an option to the `batch()` method that would be something like `raise_errors` (defaulting to `True`, I think), and when it's true then just raise the first exception that you generate. This way it'll let users get the results of all their commands even if they failed or simply get an exception raised.

I looked at the current non-legacy client and they don't handle batch operations in a particular way, so I think we may be able to contribute that code back to them, and it's not going to be for nothing. That's why I suggest doing this more complex thing.

What do you think?